lib/nerves_wpa_supplicant.ex

```diff
@@ -21,21 +21,67 @@ defmodule Nerves.WpaSupplicant do
 
   defstruct port: nil,
             ifname: nil,
+            restart: :transient,
+            control_socket_path: "",
             requests: []
 
   @doc """
   Start and link a Nerves.WpaSupplicant that uses the specified control
   socket.
   """
-  def start_link(ifname, control_socket_path, opts \\ []) do
-    GenServer.start_link(__MODULE__, {ifname, control_socket_path}, opts)
+  def start_link(ifname, control_socket_path, restart \\ :transient, opts \\ []) do
+    GenServer.start_link(__MODULE__, {ifname, control_socket_path, restart}, opts)
+  end
+  def start_link([ifname, control_socket_path, restart, opts]) do
+    start_link(ifname, control_socket_path, restart, opts)
+  end
+  def start_link([ifname, control_socket_path, opts]) do
+    start_link(ifname, control_socket_path, opts)
   end
 
-  @doc """
-  Stop the Nerves.WpaSupplicant control interface
-  """
-  def stop(pid) do
-    GenServer.stop(pid)
+  defp wait_for_disconnect(pid, ifname, timeout) do
+    Logger.info("Waiting for disconnect event for #{ifname}")
+
+    receive do
+      {Nerves.WpaSupplicant, e = {:"CTRL-EVENT-DISCONNECTED", _mac, _map}, %{ifname: ^ifname}} ->
+        Logger.debug("Got disconnect event #{inspect e}")
+        :ok
+
+      other ->
+        Logger.debug("Got event: #{inspect(other)}")
+        :timer.sleep 111
+        wait_for_disconnect(pid, ifname, timeout)
+
+      after
+        timeout ->
+          Logger.warn("Haven't received disconnect control even!")
+          :timeout
+    end
+  end
+
+   @doc """
+   Stop the Nerves.WpaSupplicant control interface
+   """
+  def stop(wpa_pid, supplicant_port \\ nil) do
+    ifname = ifname(wpa_pid)
+
+    if is_port(supplicant_port) and is_pid(wpa_pid)  do
```

Comment:
can is_pid() be used as part of the function matching clause?

Reply:
Not quite, although I have thought about it. In this particular case having if statement makes the code a bit more readable. I corrected a small error of acquiring ifname if wpa_pid = nil ;)